RFC/src/RFC-0170_NetworkCommunicationProtocol.md

```diff
@@ -0,0 +1,301 @@
+| linked asset ids | list of asset ids  | Asset ids can be used as an address on Tari network similar to a node id |
+| last_connection  | timestamp          | Time of last successful connection with peer                             |
+
+When a new CC or CN wants to join the Tari communication network they need to submit a joining request to the rest of the network.
+The joining request contains the peer address of the new CC or CN.
+Each CN that receives the joining request can decide if they want to add the new CCs or CNs contact information to their local routing table.
+When a CN, that received the joining request, has a similar node id to the new CC or CN then that node must add the peer address to their routing table.
+All CNs with similar node ids to the new CC or CN should have a copy of the new peer address in their routing tables.
+
+To limit potential attacks, only one registration with the same online communication address can be stored in the routing table of a CN.
+This restriction will limit Bad Actors from spinning up multiple CNs on a single computer.
+
+#### Joining the Network using a Joining Request
+
+A new CC or CN needs to register their peer address on the Tari communication network.
+This is achieved by submitting a network joining request to a subset of CNs selected from the routing table of the new CC or CN.
+These peers will forward the joining request, with the peer address, to the rest of the network until CNs with similar node ids have been reached.
+CNs with similar node ids will then add the new peer address of the new node to their routing table, allowing for fast discovery of the new CC or CN. 
+
+Other CCs and CNs will then be able to retrieve the new CCs or CNs peer address by submitting discovery requests.
+Once the peer address of the desired CC or CN has been discovered then a direct P2P communication channel can be established between the two parties for any future communication.
+After discovery, the rest of the Tari communication network will be unaware of any further communication between the two parties.
+
+#### Sending Data Messages and Discovery Requests
+
+The majority of all communication on the Tari communication network will be performed using direct P2P channels established between different CCs and CNs once they are aware of the peer addresses of each other that contain their online communication addresses.
+Message propagation on the network will typically consist only of joining and discovery requests where a CC or CN wants to join the network or retrieve the peer address of another CC or CN so that a direct P2P channel can be established.
+
+Messages can be transmitted in this network in either an unencrypted or encrypted form.
+Typically messages that have been sent in unencrypted form is of interest to a number of CNs on the network and should be propagated so that every CN that is interested in that data message obtains a copy.
```

Comment:
are we using something like dandelion? Or just plain old gossip?

Reply:
I believe dandelion will not be a good option in this specific case as the reason we are applying gossip is to ensure that immediately we send the message to as many peers as possible in the hope that at least one of them is good and propagate the message. We can discuss this further if needed.

Reply:
Dandelion is maybe a good idea when a wallet wants to submit a transaction to a base node.

Reply:
- If a peer connection was established between a wallet and a base node, both will know each other's online address, otherwise the connection could not be established in the first place. When the transaction is then included in the mempool of the base node and it starts propagating to other base nodes. Knowledge of the wallets online address won't be known by any of the other base nodes.
- The next example is similar to the Dandelion mechanism. If the wallet wants to keep knowledge of the transaction private, he can encrypt the transaction as a data message that is sent to some other base node through the network. Once it reaches the destination base node, it will decrypt the message and add the transaction to its mempool. The first CN that was contacted by the wallet will only know that a data message was sent, which could be anything such as a discovery request. In this case, every hop might still know the public key of the wallet so they can filter spam messages on the network.

Reply:
second option could work, but it will be upto the wallet to check the mempool to ensure it reaches the mempool. 

I know dandelion has a fall back that states every node in the stem fase must check it, if it doesnt reach the mempool they transmit.  But thats a minor thing, and not enforceable.

Reply:
Maybe adding something like this can value add?

"_Privacy of a Wallet's IP address, whom is sending a Tx to a Base Node, MAY be enhanced if the Tx is encrypted and sent to a Base Node with node ID that is not the closest to the Wallet's node ID. This should prevent linking a Tx to the originating IP address._"

Reply:
Added slightly modified version to "Sending Data Messages and Discovery Requests" section

Reply:
Looks good to me